loadtest/runner/base_load_test_runner.go

```diff
@@ -0,0 +1,670 @@
+				txToBlockMap[txn.Hash()] = receipt.BlockNumber
+			}
+		}
+	}
+
+	totalTxns := len(txHashes)
+
+	// split the txHashes into batches so we can get them in parallel routines
+	batchSize := totalTxns / 10
+	if batchSize == 0 {
+		batchSize = 1
+	}
+
+	var wg sync.WaitGroup
+
+	for i := 0; i < totalTxns; i += batchSize {
+		end := i + batchSize
+		if end > totalTxns {
+			end = totalTxns
+		}
+
+		wg.Add(1)
+
+		go func(txHashes []types.Hash) {
+			defer wg.Done()
+
+			getTxReceipts(txHashes)
+		}(txHashes[i:end])
+	}
+
+	wg.Wait()
+
+	if len(foundErrors) > 0 {
+		fmt.Println("Errors found while waiting for receipts:")
+
+		for _, err := range foundErrors {
+			fmt.Println(err)
+		}
+	}
+
+	return blockInfoMap, txnStats
+}
+
+// waitForReceipt waits for the transaction receipt of the given transaction hash.
+// It continuously checks for the receipt until it is found or the timeout is reached.
+// If the receipt is found, it returns the receipt and nil error.
+// If the timeout is reached before the receipt is found, it returns nil receipt and an error.
+func (r *BaseLoadTestRunner) waitForReceipt(txHash types.Hash) (*ethgo.Receipt, error) {
+	timer := time.NewTimer(r.cfg.ReceiptsTimeout)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			receipt, err := r.client.GetTransactionReceipt(txHash)
+			if err != nil {
+				if err.Error() != "not found" {
+					return nil, err
+				}
+			}
+
+			if receipt != nil {
+				return receipt, nil
+			}
+		case <-timer.C:
+			return nil, fmt.Errorf("timeout while waiting for transaction %s to be processed", txHash)
+		}
+	}
+}
+
+// calculateTPS calculates the transactions per second (TPS) for a given set of
+// block information and transaction statistics.
+// It takes a map of block information and an array of transaction statistics as input.
+// The function iterates over the transaction statistics and calculates the TPS for each block.
+// It also calculates the minimum and maximum TPS values, as well as the total time taken to mine the transactions.
+// The calculated TPS values are displayed in a table using the tablewriter package.
+// The function returns an error if there is any issue retrieving block information or calculating TPS.
+func (r *BaseLoadTestRunner) calculateTPS(blockInfos map[uint64]*BlockInfo, txnStats []txStats) error {
+	fmt.Println("=============================================================")
+	fmt.Println("Calculating results...")
+
+	var (
+		totalTxs        = len(txnStats)
+		totalTime       float64
+		maxTxsPerSecond float64
+		minTxsPerSecond = math.MaxFloat64
+	)
+
+	blockTimeMap := make(map[uint64]uint64)
+	uniqueBlocks := map[uint64]struct{}{}
+
+	for _, stat := range txnStats {
+		if stat.block == 0 {
+			continue
+		}
+
+		uniqueBlocks[stat.block] = struct{}{}
+	}
+
+	for block := range uniqueBlocks {
+		currentBlockTxsNum := 0
+		parentBlockNum := block - 1
```

Comment:
What will happen for block #1, parentBlock will be 0, do we need to handle that case?

Reply:
No need, the genesis or 0 block is a block as well, so it will have no transactions in it.